src/decoder.js

```diff
@@ -1,221 +1,246 @@
-import tags from './type-tags.js';
+/* global Uint8Array */
 
-/*
+import tags from './type-tags.js';
+import Extendable from './extendable.js';
+import type { ExtensionMap, ExtensionPoint } from './extendable.js';
 
-Done:
-uints
-strings
-basic values
-arrays
-maps
-barrays, bmaps
-nints
-utf-8
-float
+type SuperPackedValue = Array<number>;
+type Keyset = Array<string>;
 
+/*
 todo:
 mapl, bmapl
 timestamp
 extension
 factor out various common threads
 */
 
-let stringLUT: Array<string>, keysetLUT: Array<Array<string>>, ptr: number;
+export default class Decoder extends Extendable {
+  strings: Array<string>
+  keysets: Array<Keyset>
+  buffer : SuperPackedValue
+  ptr: number
+
+  constructor() {
+    super();
+    this.strings = [];
+    this.keysets = [];
+    this.buffer = [];
+    this.ptr = 0;
+  }
 
-export default function decode(buf: Array<number>, options?: { omittedKeysets?: Array<Array<string>> } = {}): any {
-  ptr = 0;
-  if (buf[0] === tags.STRLUT) {
-    stringLUT = (decodeValue(buf): Array<string>);
-    keysetLUT = (decodeValue(buf): Array<Array<string>>);
-  } else {
-    stringLUT = [];
-    keysetLUT = [];
+  static decode(buffer : SuperPackedValue, options? : { omittedKeysets? : Array<Keyset>, extensions? : ExtensionMap } = {}) : any {
+    let d = new Decoder;
+    if (options.extensions != null) {
+      // $FlowFixMe: flow doesn't understand that ext is an ExtensionPoint
+      Object.keys(options.extensions).forEach((ext : ExtensionPoint) => {
+        // $FlowFixMe: flow doesn't understand that options.extensions is non-null here
+        let extension = options.extensions[ext];
```

Comment:
I think you can solve this by doing `const extensions = options.extensions` before the `if`. Flow doesn't know that `options` hasn't been mutated.